server/chat-commands/room-settings.ts

```diff
@@ -1382,3 +1204,82 @@ export const commands: ChatCommands = {
 	},
 	removeroomaliashelp: [`/removeroomalias [alias] - removes the given room alias of the room the command was entered in. Requires: & ~`],
 };
+
+export const roomSettings: SettingsHandler[] = [
+	// modchat
+	(room, user) => {
+		const threshold = user.can('makeroom') ? Infinity :
+			user.can('modchatall', null, room) ? Config.groupsranking.indexOf(room.getAuth(user)) :
+			user.can('modchat', null, room) ? 1 :
+			null;
+
+		const permission = !!threshold;
+
+		// typescript seems to think that [prop, true] is of type [prop, boolean] unless we tell it explicitly
+		const options: [string, string | true][] = !permission ? [[room.modchat || 'off', true]] :
+			[
+				'off',
+				'autoconfirmed',
+				'trusted',
+				...RANKS.slice(1, threshold + 1),
+			].map(rank => [rank, (rank === 'off' ? !room.modchat : rank === room.modchat) || `modchat ${rank || 'off'}`]);
+
+		return {
+			label: "Modchat",
+			permission,
+			options,
+		};
+	},
+	(room, user) => ({
+		label: "Modjoin",
+		permission: room.isPersonal ? user.can('editroom', null, room) : user.can('makeroom'),
+		options: [
+			'off',
+			'autoconfirmed',
+			// groupchat ROs can set modjoin, but only to +
+			// first rank is for modjoin off
+			...RANKS.slice(1, room.isPersonal && !user.can('makeroom') ? 2 : undefined),
+		].map(rank =>
+			[rank, (rank === 'off' ? !room.modjoin : rank === room.modjoin) || `modjoin ${rank || 'off'}`]
+		),
+	}),
+	(room, user) => ({
+		label: "Language",
+		permission: user.can('editroom', null, room),
```

Comment:
...wonder if it's a good idea to let this return `bool | string` to save on the `user.can(` every time...

Reply:
That's a good point. I'd support that.